src/TorchSharp/Data/DataLoader.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+
+using static TorchSharp.torch;
+
+namespace TorchSharp.Data
+{
+    public class DataLoader: IEnumerable<(Tensor, Tensor)>
+    {
+        private Dataset dataset;
+        private int batchSize;
+        private bool shuffle;
+        private Device device;
+
+        public DataLoader(Dataset dataset, int batchSize, bool shuffle = false, Device device = null)
+        {
+            this.dataset = dataset;
+            this.batchSize = batchSize;
+            this.shuffle = shuffle;
+            this.device = device ?? CPU;
+        }
+        public IEnumerator<(Tensor, Tensor)> GetEnumerator()
+        {
+            return new DataLoaderEnumerator(dataset, batchSize, shuffle, device);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public long Count => (dataset.Count() - 1) / batchSize + 1;
+
+        private class DataLoaderEnumerator : IEnumerator<(Tensor, Tensor)>
+        {
+            private Dataset dataset;
+            private IEnumerator<object> load;
+            private int batchSize;
+            private Device device;
+            public DataLoaderEnumerator(Dataset dataset, int batchSize, bool shuffle, Device device)
+            {
+                this.dataset = dataset;
+                this.batchSize = batchSize;
+                this.device = device;
+
+                if (shuffle)
+                {
+                    var r = new Random();
+                    load = dataset.GetDataEnumerable()
+                        .Select(x => new {Number = r.Next(), Item = x})
+                        .OrderBy(x => x.Number)
+                        .Select(x => x.Item)
+                        .GetEnumerator();
```

Comment:
create an array of indexes from data set e.g. `Enumerable.Range(0, dataset.Count).ToArray()` then fischer yates shuffle that and keep this array of shuffled indexes as what will be iterated over, order by random is not a great way of doing shuffling. 

Fischer-Yates shuffle code is simple:
```csharp
       // Fisher-Yates shuffling
        internal static void Shuffle<T>(T[] values, Random random)
        {
            int n = values.Count;
            while (n > 1)
            {
                int k = random.Next(n);
                --n;
                T temp = values[n];
                values[n] = values[k];
                values[k] = temp;
            }
        }
```

Reply:
If not shuffling simply use indexes without shuffling, this means you can remove the GetDataEnumerable, it also fits better in potential multi-threading and pre-loading.

Reply:
There are ways of visiting every index in "random" order, without enumerating the list up front;

https://lemire.me/blog/2017/09/18/visiting-all-values-in-an-array-exactly-once-in-random-order/
https://lemire.me/blog/2017/09/26/benchmarking-algorithms-to-visit-all-values-in-an-array-in-random-order/

Something like that would be better than the Fisher-Yates shuffle, especially if you're processing very large numbers of samples.